### PR TITLE
Add Sigstore workload attestor for SPIRE

### PR DESCRIPTION
**Pull request by willallves**
**Pull Request check list**

- [ ] Commit conforms to CONTRIBUTING.md?
- [ ] Proper tests/regressions included?
- [ ] Documentation updated?

**Affected functionality**
This work extends the existing proposals to use SPIFFE/SPIRE in supply chain processes by leveraging SPIFFE/SPIRE to validate the authenticity of the workload. 
The proposal makes use of the signatures and artifacts generated during the earlier stages of the secure software supply chain process to attest to the origin of the workload. For this, we’ve chosen the Sigstore ecosystem and artifacts. Digging deeper, we are proposing using: Rekor (signature transparency log), Fulcio (signing certificate issuer and certificate transparency log) and Cosign (container image signing tool) to guarantee the authenticity of the running workload.

**Description of change**
<b style="font-weight:normal;" id="docs-internal-guid-4faab5c2-7fff-a635-8d0c-cf9c94fb8d8d"><p dir="ltr" style="line-height:1.38;margin-top:0pt;margin-bottom:0pt;"><span style="font-size:11pt;font-family:Arial;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">The proposal is to expand the existing k8s workload attestor by adding new capabilities for signature verification through the sigstore. We plan to fork the current version of k8s attestor and add the following selectors that will check every container in the pod:</span></p><br /><div dir="ltr" style="margin-left:0pt;" align="left">

Selector | Value
-- | --
k8s:containerID:image-signature-content | The value of the signature itself in a hash.
k8s:containerID:image-signature-subject | OIDC principal that signed it​
k8s:containerID:image-signature-logid | A unique LogID for the Rekor transparency log​
k8s:containerID:image-signature-integrated-time | The date when the image signature was integrated into the signature transparency log​
k8s:sigstore-validation | The confirmation if the signature is valid, has value of "passed" (eg. "k8s:sigstore-validation:passed")

</div></b>

**Which issue this PR fixes**

### CODE REVIEW

**Comment by amartinezfayo** [comment]
Thank you @willallves for this contribution!
This PR introduces changes in critical parts and needs to be thoroughly reviewed. We will be pretty busy during KubeCon EU next week, so I apologize that reviewing this will take some additional time.
Thanks again!

File: test/integration/suites/k8s-sigstore-e2e/00-setup [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+# Download kubectl at the expected version.
+download-kubectl "${KUBECTL_PATH}"
+
+# We must supply an absolute path to the configuration directory. Replace the
+# CONFDIR variable in the kind configuration with the conf directory of the 
+# running test.
+sed -i.bak "s#CONFDIR#${PWD}/conf#g" conf/kind-config.yaml
+rm conf/kind-config.yaml.bak
+
+#create folder which will contain the registry certificate information
```
Comment (MarcosDY) [inline test/integration/suites/k8s-sigstore-e2e/00-setup:22]: can you add an space after `#` in all comments? and to be consistent first word must start in uppercase

File: test/integration/suites/k8s-sigstore-e2e/00-setup [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+mkdir ${PWD}/conf/docker-registry
+mkdir ${PWD}/conf/docker-registry/certs
```
Comment (MarcosDY) [inline test/integration/suites/k8s-sigstore-e2e/00-setup:24]:
```suggestion
mkdir -p ${PWD}/conf/docker-registry/certs
```

**Review by MarcosDY — COMMENTED** [review]
Thanks for this!!!

I did an initial review, I didn't review unit test that may be refactored, based in some comments, same thing about integration tests.
I'll keep reviewing implementation.

**Comment by MarcosDY** [comment]
In order to simplify the review of all this changes, can you split this big PR into different PRs?
for example:
- PR1: Core changes
- PR2: k8s-workload-registrar changes
- PR3: integration tests

File: pkg/agent/plugin/workloadattestor/k8s/sigstore/sigstore.go [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+	}
+
+	if bundlebody.Spec.Signature.Content == "" {
+		return "", errors.New("Bundle payload body has no signature content")
+	}
+
+	return bundlebody.Spec.Signature.Content, nil
+}
+
+type SelectorsFromSignatures struct {
```
[the remark this answers is not in the document]
Comment (willallves) [reply pkg/agent/plugin/workloadattestor/k8s/sigstore/sigstore.go:211]: It is used in `package k8s` too

File: pkg/agent/plugin/workloadattestor/k8s/sigstore/sigstore.go [hunk 4]
```diff
@@ -0,0 +1,440 @@
...
+			sigstore.logger.Error("Error getting signature bundle: ", err.Error())
+		} else {
+			sigContent, err := getBundleSignatureContent(bundle)
+			if err != nil {
+				sigstore.logger.Error("Error getting signature content: ", err.Error())
+			} else {
+				selectorsFromSignatures.Content = sigContent
+			}
+			if bundle.Payload.LogID != "" {
+				selectorsFromSignatures.LogID = bundle.Payload.LogID
```
[the remark this answers is not in the document]
Comment (willallves) [reply pkg/agent/plugin/workloadattestor/k8s/sigstore/sigstore.go:257]: non-boolean condition in if statement, for this "" if LogID is ""

File: pkg/agent/plugin/workloadattestor/k8s/sigstore/sigstore.go [hunk 5]
```diff
@@ -0,0 +1,440 @@
...
+	SelectorValuesFromSignature(oci.Signature, string) SelectorsFromSignatures
+	ExtractSelectorsFromSignatures(signatures []oci.Signature, containerID string) []SelectorsFromSignatures
+	ShouldSkipImage(imageID string) (bool, error)
+	AddSkippedImage(imageID string)
+	ClearSkipList()
+	AddAllowedSubject(subject string)
+	EnableAllowSubjectList(bool)
+	ClearAllowedSubjects()
+	SetRekorURL(rekorURL string) error
+	SetLogger(logger hclog.Logger)
```
[the remark this answers is not in the document]
Comment (willallves) [reply pkg/agent/plugin/workloadattestor/k8s/sigstore/sigstore.go:44]: They are used on sigstore and k8s packages

**Pull request closed without merging**